internal/provider/meta.go

```diff
@@ -0,0 +1,399 @@
+		clientConfig.HttpClient.Transport,
+		helper.DefaultTransportOptions(),
+	)
+
+	// enable ReadYourWrites to support read-after-write on Vault Enterprise
+	clientConfig.ReadYourWrites = true
+
+	// set default MaxRetries
+	clientConfig.MaxRetries = DefaultMaxHTTPRetries
+
+	client, err := api.NewClient(clientConfig)
+	if err != nil {
+		return nil, fmt.Errorf("failed to configure Vault API: %s", err)
+	}
+
+	// setting this is critical for proper namespace handling
+	client.SetCloneHeaders(true)
+
+	// setting this is critical for proper client cloning
+	client.SetCloneToken(true)
+
+	// Set headers if provided
+	headers := d.Get("headers").([]interface{})
+	parsedHeaders := client.Headers().Clone()
+
+	if parsedHeaders == nil {
+		parsedHeaders = make(http.Header)
+	}
+
+	for _, h := range headers {
+		header := h.(map[string]interface{})
+		if name, ok := header["name"]; ok {
+			parsedHeaders.Add(name.(string), header["value"].(string))
+		}
+	}
+	client.SetHeaders(parsedHeaders)
+
+	client.SetMaxRetries(d.Get("max_retries").(int))
+
+	MaxHTTPRetriesCCC = d.Get("max_retries_ccc").(int)
+
+	// Try an get the token from the config or token helper
+	token, err := GetToken(d)
+	if err != nil {
+		return nil, err
+	}
+
+	// Attempt to use auth/<mount>login if 'auth_login' is provided in provider config
+	authLoginI := d.Get("auth_login").([]interface{})
+	if len(authLoginI) > 1 {
+		return "", fmt.Errorf("auth_login block may appear only once")
+	}
+
+	if len(authLoginI) == 1 {
+		authLogin := authLoginI[0].(map[string]interface{})
+		authLoginPath := authLogin[consts.FieldPath].(string)
+		authLoginNamespace := ""
+		if authLoginNamespaceI, ok := authLogin[consts.FieldNamespace]; ok {
+			authLoginNamespace = authLoginNamespaceI.(string)
+			client.SetNamespace(authLoginNamespace)
+		}
+		authLoginParameters := authLogin[consts.FieldParameters].(map[string]interface{})
+
+		method := authLogin[consts.FieldMethod].(string)
+		if method == "aws" {
+			logger := hclog.Default()
+			if logging.IsDebugOrHigher() {
+				logger.SetLevel(hclog.Debug)
+			} else {
+				logger.SetLevel(hclog.Error)
+			}
+			if err := signAWSLogin(authLoginParameters, logger); err != nil {
+				return nil, fmt.Errorf("error signing AWS login request: %s", err)
+			}
+		}
+
+		secret, err := client.Logical().Write(authLoginPath, authLoginParameters)
+		if err != nil {
+			return nil, err
+		}
+		token = secret.Auth.ClientToken
+	}
+	if token != "" {
+		client.SetToken(token)
+	}
+	if client.Token() == "" {
+		return nil, errors.New("no vault token found")
+	}
+
+	skipChildToken := d.Get("skip_child_token").(bool)
+	if !skipChildToken {
+		err := setChildToken(d, client)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// Set the namespace to the requested namespace, if provided
+	namespace := d.Get(consts.FieldNamespace).(string)
+	if namespace != "" {
+		client.SetNamespace(namespace)
+	}
+
+	return &ProviderMeta{
+		resourceData: d,
+		client:       client,
```

Comment:
> From the flow of the code, I can't tell where `m` the mutex is being set on the ProviderMeta; should we set that value and return it with the ProviderMeta here? I see that if `clientCache` is `nil`, we set it within the `GetNSClient` function, but it initially looks like we currently don't set the mutex to the ProviderMeta.

The `clientCache` needs to be instantiated for each `ProviderMeta` instance, since it is a map, whereas `sync.RWMutex`, a struct, does not since it is being used in the context of a struct member.

> 
> I initially suspected that this might cause a panic in `GetNSClient` when we attempt to lock the mutex, but in your test for `GetNSClient` I see that you have good test coverage on the cache locking, and since that test is passing then `m` is definitely being set somewhere. Could you help shed some light on where that's happening? :)

Since the lock is not share we don't need a pointer to it.  Each `ProviderMeta` instance will have its own copy.